### PR TITLE
Add generic typing support for Memory[Send/Receive]Channel

### LINKED ISSUE #1327: `MemoryXXXChannel` doesn't support generic typing
When using `mypy` and type hints for `trio.abc.SendChannel` and `trio.abc.ReceiveChannel`, `mypy` complains about calling things like:

- `channel.clone()`
- `channel.send_nowait()`

This is because `trio.abc.XXXChannel` don't have these interfaces, only the `MemoryXXXChannel` does.  The problem is that the `trio.abc.SendChannel` allow for specifying the type of object that passes through the channel via `trio.abc.SendChannel[SomeType]`, however, the `MemoryXXXChannel` objects do not.

Both of the solutions for this are not ideal.

- `cast(MemoryXXXChannel, my_channel)`: this loses the object type that passes through the channel.
- `# type: ignore`: this just silences the error....

### PR DESCRIPTION
**Pull request by jakkdl**
Fixes #1327

I didn't figure out a good way to add tests for it, but I reproduced the error from the original PR, and it now works as expected (in a somewhat pared down variant of it):
```python
from trio import open_memory_channel, MemorySendChannel, MemoryReceiveChannel
from typing import cast, Tuple

OutboundBroadcastChannelPair = Tuple[
    MemorySendChannel[int],
    MemoryReceiveChannel[int],
]

(outbound_send_channel, outbound_receive_channel) = cast(
    OutboundBroadcastChannelPair, open_memory_channel(100)
)

outbound_send_channel.send_nowait(5)
outbound_send_channel.send_nowait(None) # argument 1 has incompatible type

reveal_type(outbound_receive_channel.receive_nowait()) # revealed type is builtins.int
```
One way would be to add type hints to the tests in `trio/tests/test_channel.py`, and also check the `trio.tests` submodule with mypy in `check.sh`

As noted in https://github.com/python-trio/trio/blob/master/trio/_util.py#L251-L254 `@generic_function` won't type-check, and I didn't figure out what mypy plugin / clever stub would enable that.

I didn't manage to figure out what `trio.lowlevel.wait_task_rescheduled` actually returns, so it's possible that the return type on `MemoryReceiveChannel.receive` is incorrect.

I assumed that the co/contravariantness of the `TypeVar`s in `trio/_abc.py` for `SendChannel` and `MemoryChannel` would apply for Memory[Send/Receive]Channel as well.

While on it I also typed ~everything in `trio/_channel.py`. (though the `__exit__` ones look kinda crazy :sweat_smile: )

### CODE REVIEW

File: trio/_channel.py
Comment (A5rocks) [inline trio/_channel.py:140]: Should `SendChannel` be generic?

File: trio/_channel.py
Comment (A5rocks) [inline trio/_channel.py:285]: Should `ReceiveChannel` be generic?

File: trio/_channel.py
Comment (A5rocks) [inline trio/_channel.py:37]: This feels a bit wrong but thinking about it, it looks right. (I made some code examples to prove it to myself, this variance stuff is really tricky).

File: trio/_channel.py
Comment (A5rocks) [inline trio/_channel.py:201]: Should the argument be typed here?

File: trio/_channel.py
Comment (A5rocks) [inline trio/_channel.py:238]: This should return `Self`. (from `typing_extensions`, that is). (unlike `clone` above, this doesn't always return `MemorySendChannel`.)

If we don't want to use that, we can do `def __enter__(self: SelfT) -> SelfT` where `SelfT = TypeVar("SelfT")`.

File: trio/_channel.py
Comment (A5rocks) [inline trio/_channel.py:350]: Why does this need a `# type: ignore` but not the other similar pattern in `MemorySendChannel`.

File: trio/_channel.py
Comment (A5rocks) [inline trio/_channel.py:384]: Should return `Self`, I think.

**Review by A5rocks — CHANGES_REQUESTED** [review]
I've got a couple nitpicks re: types

File: trio/_channel.py
Comment (jakkdl) [reply trio/_channel.py:140]: the abc has generic, so I think you're right that the type should be passed through to it as well here.

File: trio/_channel.py
Comment (jakkdl) [reply trio/_channel.py:201]: Yes, but I gave up on figuring out what it should be :sweat_smile:

I started trying to figure out `trio.lowlevel.wait_task_rescheduled` but I had a lot of trouble with it and ultimately decided to postpone tackling lowlevel - since at least in this case it doesn't matter when the parameter is unused.
[looking at trio-typing](https://github.com/python-trio/trio-typing/blob/d5835f7065f58ebdd95ca27b09d1d5cbe686b8b8/trio-stubs/lowlevel.pyi#L131-L133) i think it should be `Callable[[], NoReturn]`, though don't ask me why.

Also see discussion further down on the `type: ignore`

File: trio/_channel.py
Comment (jakkdl) [reply trio/_channel.py:350]: `MemorySendChannel.send()` doesn't return anything, which this method does. The problem is that `wait_task_rescheduled` isn't typed so we get an `Any` as return type, which breaks us having specified the method to return `ReceiveType`.

When I looked into `trio-typing` it explicitly says that `wait_task_rescheduled` returns `Any`
```python
async def wait_task_rescheduled(
    abort_func: Callable[[Callable[[], NoReturn]], Abort]
) -> Any: ...
```
https://github.com/python-trio/trio-typing/blob/d5835f7065f58ebdd95ca27b09d1d5cbe686b8b8/trio-stubs/lowlevel.pyi#L131-L133

but I didn't manage to figure out whether that's a limitation in the typing, or what's up with that. (as I admitted in my first post).

File: trio/_channel.py
Comment (jakkdl) [reply trio/_channel.py:37]: As I said in my first post I just assumed it'd be the same as the abc, so thanks for double checking it :)

File: trio/_channel.py
Comment (A5rocks) [reply trio/_channel.py:201]: We could always just do `_: object` :^)

(in all seriousness I'll take a look and see eventually:tm:)

File: trio/_channel.py
Comment (A5rocks) [reply trio/_channel.py:350]: We could probably just move the `return` to a new line. Let's verify with `wait_task_rescheduled` first though, maybe? Perhaps this is old code.

**Comment by A5rocks** [comment]
cc @agronholm as you've had some opinions on types before

**Comment by agronholm** [comment]
The previous typing PR used `from __future__ import annotations`, so I see no reason to not do that here too. That lets you modernize the annotations too by not using deprecated types from the `typing` module.

File: trio/_channel.py
Comment (agronholm) [reply trio/_channel.py:384]: Annoyingly the latest release of mypy does not support `Self` :frowning_face: 
Support is in, it just hasn't been included in a release yet.

Comment (A5rocks) [reply trio/_channel.py:384]: Unfortunate! mypy v1.0 is right around the corner so maybe this PR could be blocked on that? Actually I don't know how soon that will be, maybe the `SelfT` hack would be best for now and we can make a PR cleaning it up soon :P

File: trio/_channel.py
Comment (jakkdl) [reply trio/_channel.py:140]: Ye, if the type doesn't get passed through then this code raises no errors on the assignment to bar:
```python
foo: MemorySendChannel[int] = ...
bar: SendChannel[str] = foo
```

File: trio/_channel.py
Comment (jakkdl) [reply trio/_channel.py:384]: Typed it as `SelfT` and added a comment saying it's temporary until mypy releases support for it.

File: trio/_channel.py
Comment (jakkdl) [reply trio/_channel.py:350]: Moved it to a new line, but yeah `wait_task_rescheduled` need to be figured out.

**Comment by jakkdl** [comment]
Pushed a commit with updates from the reviews, though I got stuck trying to resolve `open_memory_channel`. It currently complains

```python
trio/_channel.py:100: error: Incompatible return value type
(got "Tuple[MemorySendChannel[SendType], MemoryReceiveChannel[ReceiveType]]",
expected "Tuple[MemorySendChannel[T], MemoryReceiveChannel[T]]")  [return-value]
```

and I'm in general quite confused how to write that the function, with the `@generic_function` decorator is expected to take one `type`, which is passed to `MemoryChannelState`, and then we're to return the classes typed with two other typevars, presumably bound to the `type` parameter passed, but with different `covariant`/`contravariant`ness. So if anybody has an idea how to cram all that into the function header (or class header, see below), I would love to be enlightened.

I looked at trio-typing, and they [resort to making `open_memory_channel` a class:](
https://github.com/python-trio/trio-typing/blob/2bc84bab1d7f07e85e7d6f9fc95237c347cdcf55/trio-stubs/__init__.pyi#L283-L288)
```python
# written as a class so you can say open_memory_channel[int](5)
class open_memory_channel(Tuple[MemorySendChannel[T], MemoryReceiveChannel[T]]):
    def __new__(  # type: ignore[misc]  # "must return a subtype"
        cls, max_buffer_size: float
    ) -> Tuple[MemorySendChannel[T], MemoryReceiveChannel[T]]: ...
    def __init__(self, max_buffer_size: float): ...
```
which looks kinda hackish, but maybe is fine? And they don't appear to use `@generic_function` anywhere in `trio-typing`.

File: trio/_channel.py
Comment (A5rocks) [inline trio/_channel.py:142]: I don't think the `Generic` here is necessary, FWIW (same for `MemoryReceiveChannel`)

**Comment by A5rocks** [comment]
> Pushed a commit with updates from the reviews, though I got stuck trying to resolve `open_memory_channel`. It currently complains
> 
> ```python
> trio/_channel.py:100: error: Incompatible return value type
> (got "Tuple[MemorySendChannel[SendType], MemoryReceiveChannel[ReceiveType]]",
> expected "Tuple[MemorySendChannel[T], MemoryReceiveChannel[T]]")  [return-value]
> ```
> 
> and I'm in general quite confused how to write that the function, with the `@generic_function` decorator it's expected to take one `type`, which is passed to `MemoryChannelState`, and then we're to return the channel classes typed with two other typevars, presumably bound to the `type` parameter passed, but with different `covariant`/`contravariant`ness. So if anybody has an idea how to cram all that into the function header (or class header, see below), I would love to be enlightened.

I think this is because of `NoPublicConstructor`'s `_create` function. Not sure how to fix it, but hopefully that makes some stuff clearer.

**Comment by jakkdl** [comment]
Ah, that made me think of the super simple solution!
```python
    return (
        MemorySendChannel[T]._create(state),
        MemoryReceiveChannel[T]._create(state),
    )
```

I should definitely write some test cases to check that it actually works though.

**Comment by agronholm** [comment]
> Ah, that made me think of the super simple solution!
> 
> ```python
>     return (
>         MemorySendChannel[T]._create(state),
>         MemoryReceiveChannel[T]._create(state),
>     )
> ```
> 
> I should definitely write some test cases to check that stuff actually works though.

This requires Python 3.8 at minimum, just FYI.

File: trio/_channel.py
Comment (jakkdl) [reply trio/_channel.py:142]: ah of course, thanks!

**Comment by jakkdl** [comment]
> > Ah, that made me think of the super simple solution!
> > ```python
> >     return (
> >         MemorySendChannel[T]._create(state),
> >         MemoryReceiveChannel[T]._create(state),
> >     )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I should definitely write some test cases to check that stuff actually works though.
> 
> This requires Python 3.8 at minimum, just FYI.

ah, shoot

**Comment by agronholm** [comment]
> > > Ah, that made me think of the super simple solution!
> > > ```python
> > >     return (
> > >         MemorySendChannel[T]._create(state),
> > >         MemoryReceiveChannel[T]._create(state),
> > >     )
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I should definitely write some test cases to check that stuff actually works though.
> > 
> > 
> > This requires Python 3.8 at minimum, just FYI.
> 
> ah, shoot

Actually, no. The class level `__getitem__()` method was in fact introduced in Python 3.7. It's just the `collections.abc` module that gained such support in 3.8.

**Comment by jakkdl** [comment]
Yay!

Replacing the `open_memory_channel` definition with the class-based one won't work until 3.9+, but I realized I could get the best of both worlds with an `if TYPE_CHECKING` guard. (I first thought I could write a `.pyi` only for that function, but that'd override the whole file).

I also wrote some tests using https://github.com/davidfritzsche/pytest-mypy-testing to check for expected mypy error outputs, but it's kinda broken and unmaintained, and I failed to get it working with skipping on earlier versions, so not gonna try and get that working in the CI. But I did get it passing locally at least. It might be worth trying to get trio-typings testing infrastructure replicated here.

**Comment by A5rocks** [comment]
According to the docs, `wait_task_rescheduled` returns whatever `trio.lowlevel.reschedule` is passed. I think `Any` (or `object` but pragmatism > type safety) would be best.

Maybe we should brainstorm a nicer interface for this? (we could call it a one-shot channel /s :^) But we should probably just make an issue instead of making on the spot.

File: trio/_channel.py
Comment (A5rocks) [inline trio/_channel.py:112]: CI is complaining about this, heh. We can just `type: ignore` it because we know CI is wrong here (it's not run at runtime).

File: trio/_channel.py
Comment (A5rocks) [inline trio/_channel.py:373]: Looking at the code: it's not *guaranteed* to return `ReceiveType` (you could always call `reschedule` yourself and ruin trio's day...) but it's close enough: . I think the comment should be updated the reflect this, but having the `type: ignore` is aok. (Also I misunderstood earlier -- we don't need to move the `return` to a new line, I had thought `wait_task_rescheduled` just wouldn't return)

https://github.com/python-trio/trio/blob/ca850f67c84c06e79a8d5ae6ed7b24a000e8d024/trio/_channel.py#L146

File: trio/_core/_traps.py
Comment (A5rocks) [inline trio/_core/_traps.py:68]: I'm not sure about this naming or having a type alias here. But it's your call! The type alias does seem useful, even if I'm not the biggest fan of the name.

File: trio/_core/_traps.py
Comment (A5rocks) [inline trio/_core/_traps.py:70]: Discussed: if you agree with my reasoning, then the answer seems to be no, not really.

File: trio/lowlevel.py
Comment (A5rocks) [inline trio/lowlevel.py:19]: This import seems extraneous?

**Review by A5rocks — CHANGES_REQUESTED** [review]
A couple last comments, I think everything else is done.

File: trio/_channel.py
Comment (jakkdl) [reply trio/_channel.py:112]: That's because the check is run on 3.8 ... which seems kinda old by this point.
It's easily fixed by using `Tuple`, so I'll just do that until the check is run on newer stuff.

File: trio/_core/_traps.py
Comment (jakkdl) [reply trio/_core/_traps.py:70]: Feels like it might be possible to make `Task` generic and send around enough types that it all works out, but that's definitely out of scope for this PR.

File: trio/lowlevel.py
Comment (jakkdl) [reply trio/lowlevel.py:19]: Just removing it from there makes `test_core_is_properly_reexported` angry, so it's mostly a question of whether an internal `TypeVar` should be exposed in the API or not. It's not incredibly complex or anything, but might be of use to users? Unsure, I made it internal now but easy to change back.

**Comment by jakkdl** [comment]
> According to the docs, `wait_task_rescheduled` returns whatever `trio.lowlevel.reschedule` is passed. I think `Any` (or `object` but pragmatism > type safety) would be best.
> 
> Maybe we should brainstorm a nicer interface for this? (we could call it a one-shot channel /s :^) But we should probably just make an issue instead of making one on the spot.

This is definitely above my paygrade in terms of experience or opinions on trio, so a separate issue to evaluate it sounds good to me if you want to tackle it.

File: trio/lowlevel.py
Comment (A5rocks) [reply trio/lowlevel.py]: Ah didn't notice it would make a test angry, nvm then. If we're going to make a type alias we might as well go all the way and make it public

**Review by A5rocks — APPROVED** [review]
Looks good, someone else should take a look and merge if it looks good to them too.

**Review by Zac-HD — APPROVED** [review]

**Comment by jakkdl** [comment]
:rocket: 
A big thank you to @A5rocks for the very helpful and detailed reviews! :rocket:

**Pull request merged**
